air/src/constraints/mod.rs

```diff
@@ -1,108 +1,34 @@
-    /// (alpha + h2). The value h2 which is being range checked by the stack operation is stored in
-    /// the helper columns of the decoder section of the trace.
-    fn lookup_sv2(&self, alpha: E) -> E;
-    /// The value required for the fourth stack lookup when the stack requests range checks. The
-    /// value returned is the denominator used for including the value into the LogUp lookup:
-    /// (alpha + h3). The value h3 which is being range checked by the stack operation is stored in
-    /// the helper columns of the decoder section of the trace.
-    fn lookup_sv3(&self, alpha: E) -> E;
-}
+pub mod range;
 
-impl<F, E> MainFrameExt<F, E> for EvaluationFrame<F>
-where
-    F: FieldElement<BaseField = Felt>,
-    E: FieldElement<BaseField = Felt> + ExtensionOf<F>,
+/// Enforces the clock constraint: clk' = clk + 1
+///
+/// The clock must increment by 1 at each step, ensuring proper sequencing of operations.
+pub fn enforce_clock_constraint<AB>(
+    builder: &mut AB,
+    local: &MainTraceRow<AB::Var>,
+    next: &MainTraceRow<AB::Var>,
+) where
+    AB: MidenAirBuilder,
 {
-    /// Returns true when the stack operation is a u32 operation that requires range checks.
-    /// TODO: this is also defined in the op flags. It's redefined here to avoid computing all of
-    /// the op flags when this is the only one needed, but ideally this should only be defined once.
-    #[inline(always)]
-    fn u32_rc_op(&self) -> F {
-        let not_4 = binary_not(self.current()[DECODER_OP_BITS_OFFSET + 4]);
-        let not_5 = binary_not(self.current()[DECODER_OP_BITS_OFFSET + 5]);
-        self.current()[DECODER_OP_BITS_OFFSET + 6].mul(not_5).mul(not_4)
-    }
-
-    // --- Intermediate values for LogUp lookups --------------------------------------------------
-
-    #[inline(always)]
-    fn lookup_mv0(&self, alpha: E) -> E {
-        alpha + self.current()[MEMORY_D0_COL_IDX].into()
-    }
-
-    #[inline(always)]
-    fn lookup_mv1(&self, alpha: E) -> E {
-        alpha + self.current()[MEMORY_D1_COL_IDX].into()
-    }
-
-    #[inline(always)]
-    fn lookup_sv0(&self, alpha: E) -> E {
-        alpha + self.current()[DECODER_USER_OP_HELPERS_OFFSET].into()
-    }
-
-    #[inline(always)]
-    fn lookup_sv1(&self, alpha: E) -> E {
-        alpha + self.current()[DECODER_USER_OP_HELPERS_OFFSET + 1].into()
-    }
-
-    #[inline(always)]
-    fn lookup_sv2(&self, alpha: E) -> E {
-        alpha + self.current()[DECODER_USER_OP_HELPERS_OFFSET + 2].into()
-    }
-
-    #[inline(always)]
-    fn lookup_sv3(&self, alpha: E) -> E {
-        alpha + self.current()[DECODER_USER_OP_HELPERS_OFFSET + 3].into()
-    }
+    // Clock boundary constraint: clk[0] = 0
+    builder.when_first_row().assert_zero(local.clk.clone());
+
+    // Clock transition constraint: clk' = clk + 1
+    let one_expr: AB::Expr = AB::F::ONE.into();
+    builder
+        .when_transition()
+        .assert_eq(next.clk.clone(), local.clk.clone() + one_expr);
```

Comment:
Question: why do we need so many `.clone()` here? Is this how this usually works in Plonky3 (i.e., each reference to a column needs to be cloned) - or is this something specific to our implementation?